backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/ContractPartnerRepository.java

```diff
@@ -0,0 +1,34 @@
+package de.metas.contracts.bpartner;
+
+import org.adempiere.ad.dao.IQueryBL;
+import org.compiere.model.IQuery;
+import org.compiere.model.I_C_BPartner;
+import org.springframework.stereotype.Repository;
+
+import com.google.common.collect.ImmutableSet;
+
+import de.metas.bpartner.BPartnerId;
+import de.metas.contracts.model.I_C_Flatrate_Term;
+import de.metas.util.Services;
+
+@Repository
+public class ContractPartnerRepository
+{
+	public ImmutableSet<BPartnerId> retrieveAllPartnersWithContracts()
+	{
+		IQueryBL queryBL = Services.get(IQueryBL.class);
+
+		final IQuery<I_C_Flatrate_Term> contractsFilter = queryBL.createQueryBuilder(I_C_Flatrate_Term.class)
+				.addOnlyActiveRecordsFilter()
```

Comment:
now i'm wondering, what about e.g. in-progress terms. i think you need to filter only for completed and closed ones

Reply:
Right. Done.